modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/api/SessionImpl.java

```diff
@@ -543,55 +542,77 @@ List<AsyncSqlCursor<?>> openedCursors() {
     }
 
     private class ScriptHandler {
-
         private final CompletableFuture<Void> resFut;
 
+        private final AtomicReference<Throwable> cursorCloseErrHolder = new AtomicReference<>();
+
         private ScriptHandler(CompletableFuture<Void> resFut) {
             this.resFut = resFut;
         }
 
-        void processFirstResult(AsyncSqlCursor<InternalSqlRow> cursor, Throwable t) {
-            if (t != null) {
-                resFut.completeExceptionally(t);
-            } else {
-                int cursorId = registerCursor(cursor);
-                processCursor(cursor, cursorId);
+        void processCursor(AsyncSqlCursor<InternalSqlRow> cursor, Throwable scriptError) {
+            if (scriptError != null) {
+                // Stop script execution.
+                completeScriptExecution(scriptError);
+
+                return;
             }
+
+            cursor.closeAsync().whenComplete((ignored, cursorCloseError) -> {
+                if (cursorCloseError != null) {
+                    // Just save the error for later and continue fetching cursors.
+                    saveError(cursorCloseError);
+                }
+
+                if (!busyLock.enterBusy()) {
+                    completeScriptExecution(sessionIsClosedException());
+                    return;
+                }
+
+                try {
+                    if (cursor.hasNextResult()) {
+                        cursor.nextResult().whenComplete(this::processCursor);
```

Comment:
Let's avoid synchronous continuation execution under busylock.
```suggestion
                        cursor.nextResult().whenCompleteAsync(this::processCursor, pool);
```

Reply:
Can you suggest what pool should be used? Currently I decided to keep system pool.